### PR TITLE
[AIRFLOW-1557] Backfill should respect pool limit

### PR DESCRIPTION
**Pull request by milton0825**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-1557
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
The number of backfill tasks running in parallel should be limited by the open slots in pool. If no pool is specified, the number of backfill tasks per backfill should be limited by `non_pooled_backfill_task_slot_count`

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by milton0825** [comment]
PTAL @feng-tao @Fokko 
Tests passed on: 
https://travis-ci.org/milton0825/airflow/builds/509546989

File: airflow/config_templates/default_airflow.cfg [hunk 1]
```diff
@@ -125,6 +125,10 @@ dags_are_paused_at_creation = True
 # whose size is guided by this config element
 non_pooled_task_slot_count = 128
 
+# When not using pools, the number of backfill tasks per backfill
+# is limited by this config element
+non_pooled_backfill_task_slot_count = 128
```
Comment (feng-tao) [inline airflow/config_templates/default_airflow.cfg]: is there a reason not to reuse the non_pooled_task_slot_count  config?

File: airflow/jobs.py [hunk 2]
```diff
@@ -2247,6 +2247,7 @@ def _per_task_process(task, key, ti, session=None):
                             self.log.debug('Sending %s to executor', ti)
                             # Skip scheduled state, we are executing immediately
                             ti.state = State.QUEUED
+                            ti.pool = pool_name
```
Comment (feng-tao) [inline airflow/jobs.py]: will pool_name be initialized correctly?

File: tests/test_jobs.py [hunk 3]
```diff
@@ -239,9 +239,9 @@ def test_backfill_examples(self):
                 ignore_first_depends_on_past=True)
             job.run()
 
-    def test_backfill_conf(self):
+    def _get_dummy_dag(self, dag_id):
```
Comment (feng-tao) [inline tests/test_jobs.py]: since this is not a test file, maybe move it to the top of the class?

**Review by feng-tao — COMMENTED** [review]
I think overall LGTM, have you tested using backfill with celeryExecutor?

**Comment by feng-tao** [comment]
is your description correct? 
```The number of backfill tasks should be limited by the pool slots if we specified the
pool to use with:
airflow backfill --pool 
If no pool is specified, the number of backfill tasks per backfill should be limited by non_pooled_backfill_task_slot_count
```

Each backfill Dag run shared the same TI with scheduler DAGRun which should have the pool config, why do we need to provide the pool name for backfill?

File: airflow/jobs.py (hunk 2 above)
Comment (milton0825) [reply airflow/jobs.py]: yes it is initialized in:
https://github.com/apache/airflow/blob/d9adf1b4a53b07701572f27e2293f23dc8560a10/airflow/jobs.py#L2304

File: airflow/config_templates/default_airflow.cfg (hunk 1 above)
Comment (milton0825) [reply airflow/config_templates/default_airflow.cfg]: Usually people don't run backfills on the machine where scheduler runs. The backfill machine can be beefier than the scheduler machine. Thus I think it make sense to allow a larger task_slot_count for backfills and that is the reason why I make it another config.

**Comment by milton0825** [comment]
@feng-tao  
The description is correct. The backfill should either be limited by the default `non_pooled_backfill_task_slot_count` or the pool specified through `backfill --pool`. 
I think we should use a different pool for backfill instead of re-using the pool used by the normal scheduling. We can prevent backfill tasks from occupying all the pool slots and starve the normal tasks.

File: tests/test_jobs.py (hunk 3 above)
Comment (milton0825) [reply tests/test_jobs.py]: Moved to top of class

**Comment by feng-tao** [comment]
@milton0825 I think you are assuming backfill is using LocalExecutor which is only for certain use case(like Lyft). But I think there are many other use cases which use the default executor as the scheduler job(celery) which I think using the original task instance pool make senses.

WDYT @milton0825 @ashb @astahlman

**Comment by milton0825** [comment]
@feng-tao even if we are using the same executor as scheduler, I still think we should not use the same task instance pool. It is possible that the user launch a huge backfill that occupies the pool and causes normal tasks to starve. In general, I think we should have better isolation between normal tasks and backfill tasks. I feel like backfill tasks should have their own pool.

**Comment by ashb** [comment]
> It is possible that the user launch a huge backfill that occupies the pool and causes normal tasks to starve

That is the behaviour I _want_. The reason for having a pool is to limit access to an external API with rate limits. If backfill uses it's "own" pool then you have doubled the capacity of the pool.

File: airflow/config_templates/default_airflow.cfg (hunk 1 above)
Comment (ashb) [reply airflow/config_templates/default_airflow.cfg]: > Usually people don't run backfills on the machine where scheduler runs

That is a very bold, unsubstantiated claim. And one that I disagree with.

Comment (ashb) [reply airflow/config_templates/default_airflow.cfg]: Additionally if we want this to be the "same" as the setting above even when that changes we should do this 
```suggestion
non_pooled_backfill_task_slot_count = %(non_pooled_task_slot_count)s
```
And the config parser will set it to the same value.

File: airflow/config_templates/default_test.cfg [hunk 4]
```diff
@@ -48,6 +48,7 @@ dag_concurrency = 16
 dags_are_paused_at_creation = False
 fernet_key = {FERNET_KEY}
 non_pooled_task_slot_count = 128
+non_pooled_backfill_task_slot_count = 128
```
Comment (ashb) [inline airflow/config_templates/default_test.cfg]:
```suggestion
```

Not needed due to how config loading works.

File: airflow/jobs.py (hunk 2 above)
Comment (ashb) [reply airflow/jobs.py]: That line is _after_ this one.  Oh this is in a function. NM.

File: airflow/jobs.py [hunk 5]
```diff
@@ -2300,8 +2301,33 @@ def _per_task_process(task, key, ti, session=None):
...
+
             for task in self.dag.topological_sort():
                 for key, ti in list(ti_status.to_run.items()):
+                    if open_slots <= 0:
+                        self.log.debug(
+                            "Not scheduling since there are %s open slots in pool %s",
+                            open_slots, pool_name,
+                        )
+                        break
+                    open_slots -= 1
```
Comment (ashb) [inline airflow/jobs.py]: I feel that just keeping a count in memory isn't right, and we should check the DB for an accurate count of the pool each loop.

I also don't think that breaking out of the loop is the right behaviour - this would mean that if you hit the pool limit backfilling just stops, rather than waiting for a slot to become free.

**Review by ashb — CHANGES_REQUESTED** [review]
Validate assumptions about use of pools.

Don't stop the backfill when pool limit is reached, but wait for slots to be free.

File: airflow/config_templates/default_airflow.cfg (hunk 1 above)
Comment (milton0825) [reply airflow/config_templates/default_airflow.cfg]: @ashb Let me rephrase it a bit. In some cases, people do not run backfills on where scheduler runs...

Anyway I think we should still add this config so that people can fine tune the slot count if they launch backfill job on a machine other than scheduler

**Comment by milton0825** [comment]
@ashb thanks for the review! I think your concerns are totally valid and I will address those.

**Comment by feng-tao** [comment]
@milton0825 , I kinda agree with @ashb . I think pooling is to use for limit resources for different teams which having a different pool for schedule and backfill job will be confusing. This will cause the issue that doubling the capacity  as @ashb mentioned.

File: airflow/jobs.py (hunk 5 above)
Comment (milton0825) [reply airflow/jobs.py]: Backfilling would not stop actually. We will wait here for later retry:
https://github.com/apache/airflow/blob/84a7a9e6cb81139a2b62996acbd8579aff576810/airflow/jobs.py#L2484-L2508

File: airflow/config_templates/default_test.cfg (hunk 4 above)
Comment (milton0825) [reply airflow/config_templates/default_test.cfg]: fixed.

File: airflow/jobs.py (hunk 5 above)
Comment (milton0825) [reply airflow/jobs.py]: Changed to check open slots for each task

**Comment by milton0825** [comment]
@ashb @feng-tao ready for another look

File: airflow/jobs.py [hunk 6]
```diff
@@ -2300,9 +2297,35 @@ def _per_task_process(task, key, ti, session=None):
...
+                        if task.pool:
+                            pool = session.query(models.Pool) \
+                                .filter(models.Pool.pool == task.pool) \
+                                .first()
+                            if not pool:
+                                raise PoolNotFound('Unknown pool: {}'.format(task.pool))
+
+                            open_slots = pool.open_slots(session=session)
+                            if open_slots <= 0:
+                                raise NoAvailablePoolSlot(
```
Comment (feng-tao) [inline airflow/jobs.py:2316]: so now if there is no enough pool, the backfill will fail or retry?

**Review by feng-tao — APPROVED** [review]
I think overall lgtm, could wait @ashb for final comment.

File: airflow/jobs.py (hunk 6 above)
Comment (milton0825) [reply airflow/jobs.py:2316]: Whenever there is no available slots, backfill would sleep and retry.

Comment (milton0825) [reply airflow/jobs.py:2316]: FYI. This is where retry happens:
https://github.com/apache/airflow/blob/84a7a9e6cb81139a2b62996acbd8579aff576810/airflow/jobs.py#L2484-L2508

Comment (feng-tao) [reply airflow/jobs.py:2316]: thanks. lgtm

**Comment by sosso** [comment]
@feng-tao @ashb @milton0825 Just for clarification, it seems like this doesn't make any special exceptions for the 'mark success' (skip) functionality in backfill.  Is that correct?

**Comment by milton0825** [comment]
@sosso you mean you want to `mark success` some tasks in the middle of the backfill to skip those tasks?
